### DISCUSSION TITLE
Adopting of MikroORM in a large scale application

### DISCUSSION THREAD

**Discussion by klausXR** [mikro-orm/mikro-orm#2416] [General]
I have been tracking the progress of this package almost since the start. Since I started learning more about backend architecture patterns, I was rather disappointed to see that a lot of the patterns used in other languages are hardly applicable in Node and one of the main problems was the fact that there was no good choice of an ORM following the DataMapper pattern. Without that, implementing business logic in entities  instead of services becomes a lot harder and involves a lot more boilerplate. 

As of lately I have been rather happy with the progress that you are making and also happy to see that there are people that have started contributing to the project and I'm thinking about possibly adopting it into the application that I'm working on.

Since Sequelize doesn't really deliver a good experience when working with hundreds of models, a migration to something more ergonomic is imminent  at some point, and MikroORM is the only alternative, which seems promising as of now. 

I'm planning to begin with an incremental adoption. We are currently using Sequelize and will likely use both MikroORM and Sequelize during the migration period. 

Since the application is large and changes like these are time consuming and not that easy to implement, it would be in our best interest to make decisions carefully.

In that regard, I would like to ask you about any known, as of now, issues, bugs or flexibility limitations that may be problematic in larger apps, including performance. 

If some specifics have been discussed or you have something in mind,  I'd like to learn about it and where I can track its progress.

Thank you for the work that you have been doing.

This is also not addressed toward the official maintainers. If you are actually using MikroORM in a production application, I would really like to hear about the problems and hiccups that you were or are still having.

**Comment by B4nan** [comment]
Let's pin this to get more attraction, also [shared it on twitter](https://twitter.com/MikroOrm/status/1459860018820522000). Maybe @parisholley could give you some feedback, as he is adopting the library these days, and given the amount of issues he created, it looks like it is quite a large app :] You can also see some feedback in #296. I am not aware of anything really blocking/problematic, granted there are people complaining about things here and there (mostly those that never worked with data mappers in other languages), but usually it's nothing critical. 

Main issue I could see is schema diffing in v4, that lacks a lot of features and often produces additional queries (as it fails to compare some things properly) - but that is already addressed in v5.

**Comment by parisholley** [comment]
I am still in the process of refactoring a GraphQL heavy application from TypeORM to MikroORM with about ~100 database tables and used just about every feature in both (except schema/migration support in Mikro). Here is my high-level criteria as to why you should choose MikroORM (or any Unit of Work style ORM) vs TypeORM/Sequalize/etc

## Evaluating Solutions

### Complexity
Simply put, TypeORM/Sequelize, those are "stateless" libraries. If you have the hydrated entity in a variable and you save it or otherwise tweak it, you can think of it as normal javascript and queries. A UOW approach is going to add another layer of complexity and if you are in a larger team, the likelihood of other engineers having experienced this type of library is very slim (UOW is a very enterprise concept mostly used in Java, .NET, etc).

Most of the issues I've submitted to the project thus far are to help mitigate the risk associated with developers making mistakes.

### Cleanliness
While a non-UOW library will be simpler to write code and understand, you still end up effectively writing your own UOW implementation without realizing it. Take this example below:

```typescript
async function callOne(em:EntityManager, author:Author){
  author.name = 'Foobar';

 await em.save(author);
}

async function callTwo(em:EntityManager, authorId: string){
  const author = await em.getById(Author, authorId);
  author.name = 'James';

 await em.save(author);
}

async function delegate(em:EntityManager, authorId:string){
  const author = await em.getById(Author, authorId);

  await callOne(em, author);
 
  await callTwo(em, authorId);

 await em.save(author);
}
```

When you have multiple developers working on the same code base, it is easier to have one author create callOne and another create callTwo, with someone eventually needing to reference both of those inside some delegate/higher-level service call. You can see from that example that with a non-UOW work, ```callTwo``` uses a different object in memory. As soon as we save the initial `author` entity at the end of the `delegate`, the data callTwo will be lost. 

Granted, this example is rather plain/simple but the problem is exponential in a 100K line project. You effectively upfront have to commit to design patterns ahead of time:

- **A:** Make every service call only accept an id, that way you guarantee you are mutating the latest value from the database. The huge downside to this is performance, you are drastically increasing the read/write on requests.
- **B:** Pass a hydrated entity down each layer. While this will save you on select queries, every layer cannot assume that a hydrated entity would be persisted, so you have to always call `em.save()` if you intend to mutate, re-introducing write performance loss

Now if your app is simple and most of your database I/O happens in controllers without any real concept of a service layer, then this point doesn't matter.

### Performance
Personally, this is probably the only reason why I would recommend UOW style libraries. If most of your mutations/controllers are at most traversing one-level down in a model (eg: author > books) or doing one-to-one inner joins, you aren't going to see much of a performance gain. Soon as you introduce mutli-level GraphQL queries or N+1 business logic, non-UOW libraries are simply going to bring your request response times to a crawl.

https://github.com/typeorm/typeorm/issues/3857#issuecomment-918827358

The issue above has been around for years and goes to show the prioritization of performance related issues in that project. This one particular problem just eats away at CPU time which in a single-threaded Node app is crucial. Most of my performances problems isn't queries, its inefficient library code.

### Schema/Migrations
Just don't :)

TypeORM was full of bugs in generating diffs and its a game of whack-a-mole that no project will be able to solve effectively. Save pulling your hair out and just write SQL migrations your self. Ultimately, your going to need to do some database/dialect specific things at some point and it is easier to just get comfortable doing it manually.

## Why I Like MikroORM

**Note:** These examples are pseudo code, API/names may be different

### Ergonomic Lazy Loading

Mikro has a concept of "wrapped" types like `Collection` and `IdentifiedReference`, other libraries either force you to "type" your entities with the raw type (`authors: Authors[]`) or use promises (`authors: Promise<Authors[]>`). This can get really messy, really fast and the lack of explicitness (eg: is this promise resolved?, is the array empty because it wasn't eagerly loaded?) is bound to cause bugs.

### "Loaded" Entities

```typescript
async function callOne(books:Loaded<Book, 'author'>[]){
}

async function callTwo(books::Loaded<Book, 'genre'>[]){
}

async function delegate(em:EntityManager){
  const books = await em.find(Book, {populate: ['author','genre']});

  await callOne(books);
 
  await callTwo(books);
}
```

In the above call, I have two service methods which have their own data requirements. By using `Loaded`, I can enforce the consuming service/controller to do a more efficient populate query higher up in the stack before passing down the hydrated objects. This can help reduce a lot of N+1 issues and you get the benefit of complie-time checks for eager relationship loading.

### Flush Per Request

Every time you call `em.persist(entity)`, it just puts it in a write queue. At the end of the request, either via middleware or some other method, you can call `em.flush()` and it will try to send your write queries in batch, saving on I/O.

### No Mutation Risk

In my first example, you can switch between passing ids or entities around without any risk of destroying transient data. This becomes more apparent when you have references to different entity relations (one-to-many), where you are constantly adding or removing elements in various service layers.

## Why Not Mikro?

- You have a very small number of models (database tables)
- Most of your requests don't need to query many levels deep
- You have no concept of a service layer
- You don't anticipate many concurrent users

If any of those are true, I would stick to another ORM. Otherwise, I think this project is a good enough state that I would not be concerned of any production-level issues for most projects.

**Reply by rubiin** [reply]
@parisholley can I turn this on a blog post. seems like a good read

**Comment by klausXR** [comment]
Okay. I already anticipated most of these and I'm very glad that there aren't any integration issues that are connected to MikroORM ( most of the issues you mentioned are pros cons of DM/AR ).

I will wait for the next major version and will experimenting with the migration.

Is there a place you'd recommend where the ins and outs of the data mapper pattern are explained, because I see a lot of unfamiliar terms - not here, but in general, like persisting, flushing, auto-flushing at the end of a request? I do not feel confident in the understanding of how exactly the library works and I assume that using it that way will inevitably lead to some bugs or performance issues, so I'd like to familiarize myself as much as possible before diving in.

Thanks a lot, both for the explanation and for taking a leap of faith implementing it into your project

**Comment by parisholley** [comment]
> Is there a place you'd recommend where the ins and outs of the data mapper pattern are explained, because I see a lot of unfamiliar terms - not here, but in general, like persisting, flushing, auto-flushing at the end of a request?

your best bet is to look for information around Hibernate or JPA in enterprise Java as that is where a lot of these concepts originated (i believe). in particular, the concept of `first-level`/`second-level` caching like so:

https://vladmihalcea.com/jpa-hibernate-first-level-cache/

**Comment by leonardbinet** [comment]
We've been progressively migrating an application with ~100 tables from typeorm to mikro-orm, it's been going really well so far.
There is no possible comparison between those 2 orms.

Mikro-orm seems much more robust, is well documented, provides really nice features, among which:
- type safe relations 
- smart nested populate (that doesn't break the db when populating lots of relations)
- clean way of handling [collections](https://mikro-orm.io/docs/collections)

So far we've not encountered any real issue (whereas on typeorm we encountered so many weird bugs 😅 ).

On top of that, @B4nan is really very reactive on slack.

The only downside I see is the fact that this library relies on a [single person](https://github.com/mikro-orm/mikro-orm/graphs/contributors), which represents a risk.

**Reply by KrisLau** [reply]
Not a mikro ORM user rn, but thinking of migrating from objectionjs and yeah the single contributor reliance is my biggest worry after what happened with objection 😅

**Reply by rubiin** [reply]
I have been using mikroorm since like I heard it like 3- 4 yrs ago and this hasnt been a stopping factor for me. Many OSS projects have a single person incharge

**Reply by nicomouss** [reply]
@leonardbinet How went your migration from typeorm to mikroorm ? It would be interesting to read your feedback about the process, and how things are going now with the new orm.

I personally started out with typeorm, switched to mikroorm, and never went back.

Typeorm is actually more like a DataMapper, whereas Mikroorm is more like a real full-featured orm in line with Hibernate and NHibernate philosophy. Technically speaking, Mikroorm is definitely more robust, I can't agree more. No bugs, it just works. Not the same with typeorm, where you often wonder what is the next weird thing you gonna hit sooner or later.

Any feedback like yours might definitely help in the adoption of mikroorm. Would be happy to read about it !

**Reply by leonardbinet** [reply]
Hi @nicomouss, technically it went quite smoothly, it was more of an organisational challenge:
- making sure everybody understands the identity map concept
- specifying clear patterns in our codebase (since mikro-orm offers more flexibility)
- breaking up the migration in small chunks
- pushing the migration forwards during nearly 1year

The biggest technical challenges were:
- making co-exist both mikro & typeorm version of lots entities (because of a highly relational model), until every usecase stopped using typeorm entities
- making sure typeorm and mikro-orm entities were in sync with the current database schema
- implementing postgres row-level-security, within nestjs DI

As a whole, mikro-orm is a solid foundation on which we can rely (whereas typeorm sometimes felt like russian roulette), congrats @B4nan for your work 🙌

**Reply by B4nan** [reply]
> implementing postgres row-level-security, within nestjs DI

I keep getting questions about RLS in postgres, I am curious how did you end up implementing this (and I'd welcome suggestions on how to make it simpler/better, ideally we'd end up with some guide in the docs).

**Reply by leonardbinet** [reply]
I'll try to dedicate some time to answer: the short answer, we rely on:
- https://docs.nestjs.com/recipes/async-local-storage, so that each incoming request gets its own isolated context, that stores the  `tenantId`
- a connection manager, stores a mikro-orm instance (MikroORM.init) per tenant
- a tenant-scoped-entity-manager provider, relies on a factory to generate a fork for each incoming request, based on the `tenantId` stored in the async local storage (if the connection manager has a connection for this tenant, simply fork it, else create a new instance). The tenantId is passed in the postgres username: `tenant_23` -> and a policy relies on the suffix (see [more details here](https://github.com/mikro-orm/mikro-orm/discussions/6413#discussioncomment-12185530))

Notes:
- this approach works well because we serve a limited number of large companies.
With thousands of tenants, it might be an issue, and this [approach](https://github.com/mikro-orm/mikro-orm/discussions/6137#discussioncomment-11609219), based on passing the tenant as a parameter would be more suited (with the disadvantage of having to wrap all incoming requests in a transaction, which might not be suited for all applications)
- we initially relied on nestjs request-scoped providers, but ended up using async local storage for performance reasons
- we don't rely on @mikro-orm/nestjs package

**Reply by simonbrunel** [reply]
@leonardbinet Thanks for the detailed explanations, I really appreciate you taking the time to share.

Are you still satisfied with your solution after this time? Have you encountered any cases where data could potentially leak between tenants?

> We don't rely on @mikro-orm/nestjs package

Is there a specific reason for that choice? Is it mainly because you need stricter control over connection acquisition to correctly inject or reset the RLS context?

Also, do you think that [this proposed feature in v7 via Kysely](https://github.com/mikro-orm/mikro-orm/discussions/6137#discussioncomment-16625410) would make your solution significantly easier or more robust?
